### PR TITLE
Feature/plotting grid

### PR DESCRIPTION
**Pull request by amandaher**
FDataGrid ploting extended to more than one dimension on the image and up to two dimension on the domain.

### CODE REVIEW

File: examples/plot_extrapolation.py [hunk 1]
```diff
@@ -52,7 +52,7 @@
 
 # Plot of diferent representations
 fig, ax = plt.subplots(2,2)
-fdgrid.plot(ax[0][0])
+fdgrid.plot(ax = [ax[0][0]])
```
Comment (vnmabus) [inline examples/plot_extrapolation.py]: As of PEP8, named parameters are used without spaces:
```suggestion
fdgrid.plot(ax=[ax[0][0]])
```

File: examples/plot_extrapolation.py [hunk 2]
```diff
@@ -96,7 +96,7 @@
 
 
 # Plot objects in the domain range
-fdgrid.plot(ax[0][0])
+fdgrid.plot(ax = [ax[0][0]])
```
Comment (vnmabus) [inline examples/plot_extrapolation.py]:
```suggestion
fdgrid.plot(ax=[ax[0][0]])
```

File: docs/Makefile [hunk 3]
```diff
@@ -51,7 +51,7 @@ help:
 clean:
 	rm -rf $(BUILDDIR)/*
 	rm -rf auto_examples
-	rm -rf autosummary
+	rm -rf modules/autosummary
```
Comment (vnmabus) [inline docs/Makefile]: Why does this change?

File: fda/grid.py [hunk 4]
```diff
@@ -677,92 +685,157 @@ def concatenate(self, other):
                                                         other.data_matrix),
                                                        axis=0))
 
-    def _set_labels(self, ax):
+    def set_figure_and_axes(self):
```
Comment (vnmabus) [inline fda/grid.py]: Why is this public?

File: fda/grid.py [hunk 5]
```diff
@@ -677,92 +685,157 @@ def concatenate(self, other):
...
+
+        """
+        fig = plt.gcf()
+
+        if self.ndim_domain == 1:
+            projection = None
+        else:
+            projection = '3d'
+
+        ncols = math.ceil(math.sqrt(self.ndim_image))
```
Comment (vnmabus) [inline fda/grid.py]: Why the sqrt? I think I am missing something.

File: fda/grid.py [hunk 6]
```diff
@@ -677,92 +685,157 @@ def concatenate(self, other):
...
+            plt.subplots_adjust(wspace=0.4)
+
+        if nrows > 1 and self.axes_labels is not None:
+            plt.subplots_adjust(hspace=0.4)
+
+        ax = fig.get_axes()
+
+        return fig, ax
+
+    def set_labels(self, fig = None, ax = None):
```
Comment (vnmabus) [inline fda/grid.py]: Why is this public?

Comment (vnmabus) [inline fda/grid.py]:
```suggestion
    def set_labels(self, fig=None, ax=None):
```

File: fda/grid.py [hunk 7]
```diff
@@ -677,92 +685,157 @@ def concatenate(self, other):
...
+                for i in range(self.ndim_image):
+                    ax[i].set_xlabel(self.axes_labels[0])
+                    ax[i].set_ylabel(self.axes_labels[1])
+                    ax[i].set_zlabel(self.axes_labels[i + 2])
+            else:
+                for i in range(self.ndim_image):
+                    ax[i].set_xlabel(self.axes_labels[0])
+                    ax[i].set_ylabel(self.axes_labels[i + 1])
+
+    def plot(self, fig=None, ax = None, **kwargs):
```
Comment (vnmabus) [inline fda/grid.py]:
```suggestion
    def plot(self, fig=None, ax=None, **kwargs):
```

File: fda/grid.py [hunk 8]
```diff
@@ -677,92 +685,157 @@ def concatenate(self, other):
...
-        return _plot
-
-    def scatter(self, ax=None, **kwargs):
+        if self.ndim_domain > 2:
+            raise NotImplementedError("Plot only supported for functional data"
+                                      "modeled in at most 3 dimensions.")
+        if fig is not None and ax is not None:
+            raise ValueError("fig and axes parameters cannot be passed as arguments at the same time.")
+
+        if fig != None and len(fig.get_axes()) != self.ndim_image:
```
Comment (vnmabus) [inline fda/grid.py]:
```suggestion
        if fig is not None and len(fig.get_axes()) != self.ndim_image:
```

File: fda/grid.py [hunk 9]
```diff
@@ -677,92 +685,157 @@ def concatenate(self, other):
...
+            raise ValueError("Number of axes of the figure must be equal to"
+                             "the dimension of the image.")
+
+        if ax is not None and len(ax)!= self.ndim_image:
+            raise ValueError("Number of axes must be equal to the dimension of the image.")
+
+        if fig == None and ax == None:
```
Comment (vnmabus) [inline fda/grid.py]:
```suggestion
        if fig is None and ax is None:
```

File: fda/grid.py [hunk 10]
```diff
@@ -677,92 +685,157 @@ def concatenate(self, other):
...
+            fig, ax = self.set_figure_and_axes()
+
+        if self.ndim_domain == 1:
+            for i in range(self.ndim_image):
+                ax[i].plot(self.sample_points[0], self.data_matrix[:, :, i].T, **kwargs)
```
Comment (vnmabus) [inline fda/grid.py]: What happens here if you pass fig?

File: fda/grid.py [hunk 11]
```diff
@@ -677,92 +685,157 @@ def concatenate(self, other):
...
-                               **kwargs)
-        self._set_labels(ax)
-        return _plot
+        if self.ndim_domain > 2:
+            raise NotImplementedError("Plot only supported for functional data"
+                                      "modeled in at most 3 dimensions.")
+        if fig is not None and ax is not None:
+            raise ValueError("fig and axes parameters cannot be passed as arguments at the same time.")
+
+        if fig != None and len(fig.get_axes()) != self.ndim_image:
```
Comment (vnmabus) [inline fda/grid.py]:
```suggestion
        if fig is not None and len(fig.get_axes()) != self.ndim_image:
```

File: fda/grid.py [hunk 12]
```diff
@@ -677,92 +685,157 @@ def concatenate(self, other):
...
+            raise ValueError("Number of axes of the figure must be equal to"
+                             "the dimension of the image.")
+
+        if ax is not None and len(ax)!= self.ndim_image:
+            raise ValueError("Number of axes must be equal to the dimension of the image.")
+
+        if fig == None and ax == None:
```
Comment (vnmabus) [inline fda/grid.py]:
```suggestion
        if fig is None and ax is None:
```

File: fda/grid.py [hunk 13]
```diff
@@ -677,92 +685,157 @@ def concatenate(self, other):
...
-        if ax is None:
-            ax = matplotlib.pyplot.gca()
-        _plot = None
-        for i in range(self.nsamples):
-            _plot = ax.scatter(self.sample_points[0],
-                               self.data_matrix[i],
-                               **kwargs)
-        self._set_labels(ax)
-        return _plot
+        if self.ndim_domain > 2:
```
Comment (vnmabus) [inline fda/grid.py]: Not very important, but maybe the checks can be extracted to a function, as they are used twice (in `plot` and `scatter`).

In fact, the two functions are nearly identical, so maybe you can use a private function and parameterize the parts that are not common.

File: requirements.txt [hunk 14]
```diff
@@ -1,4 +1,5 @@
 matplotlib
 numpy
 scipy
-setuptools
\ No newline at end of file
+setuptools
+sklearn
```
Comment (vnmabus) [inline requirements.txt]: Technically you do not use Scikit-learn in this PR, but let it be as we will need it later.

File: docs/Makefile (hunk 3 above)
Comment (amandaher) [reply docs/Makefile]: Because the directory autosummary is generated inside modules directory.

File: fda/grid.py (hunk 4 above)
Comment (amandaher) [reply fda/grid.py]: It is also used in boxplot and MS-plot.

File: fda/grid.py (hunk 5 above)
Comment (amandaher) [reply fda/grid.py]: ´ncols = math.ceil(math.sqrt(self.ndim_image))
nrows = math.ceil(self.ndim_image / ncols)´
The figure has as many axes as dimensions on the image. Imagine there are 4 dimensions,
we want a two graphs on top of another two, not 4 in line.

File: fda/grid.py (hunk 4 above)
Comment (vnmabus) [reply fda/grid.py]: Then maybe it should not be a method, but a function in the plotting module.

File: docs/Makefile (hunk 3 above)
Comment (vnmabus) [reply docs/Makefile]: Ok, so this corrects a previous bug, in that case, because you did not change the generation.

File: fda/grid.py (hunk 5 above)
Comment (vnmabus) [reply fda/grid.py]: I do not agree. I think that the disposition should be parametrizable, at least, even if we preserve the default of making square grids.

Comment (amandaher) [reply fda/grid.py]: changed

File: fda/grid.py (hunk 10 above)
Comment (amandaher) [reply fda/grid.py]: changed

File: fda/grid.py (hunk 13 above)
Comment (amandaher) [reply fda/grid.py]: unification of checks done

File: fda/grid.py [hunk 15]
```diff
@@ -700,22 +706,24 @@ def set_figure_and_axes(self):
         else:
             projection = '3d'
 
-        ncols = math.ceil(math.sqrt(self.ndim_image))
-        nrows = math.ceil(self.ndim_image / ncols)
+        if ncols is None and nrows is None:
```
Comment (vnmabus) [inline fda/grid.py]: If only one of those is `None`, the other can be inferred.

File: fda/grid.py [hunk 16]
```diff
@@ -677,92 +685,196 @@ def concatenate(self, other):
...
+                for i in range(self.ndim_image):
+                    ax[i].set_xlabel(self.axes_labels[0])
+                    ax[i].set_ylabel(self.axes_labels[1])
+                    ax[i].set_zlabel(self.axes_labels[i + 2])
+            else:
+                for i in range(self.ndim_image):
+                    ax[i].set_xlabel(self.axes_labels[0])
+                    ax[i].set_ylabel(self.axes_labels[i + 1])
+
+    def _genereic_plotting_checks(self, fig=None, ax=None, nrows=None, ncols=None):
```
Comment (vnmabus) [inline fda/grid.py]:
```suggestion
    def _generic_plotting_checks(self, fig=None, ax=None, nrows=None, ncols=None):
```

File: fda/grid.py [hunk 17]
```diff
@@ -677,92 +685,196 @@ def concatenate(self, other):
...
-        _plot = ax.plot(self.sample_points[0],
-                        numpy.transpose(numpy.squeeze(self.data_matrix,
-                                                      axis=2)),
-                        **kwargs)
-        self._set_labels(ax)
-
-        return _plot
-
-    def scatter(self, ax=None, **kwargs):
+        fig, ax = self._genereic_plotting_checks(fig, ax, nrows, ncols)
```
Comment (vnmabus) [inline fda/grid.py]:
```suggestion
        fig, ax = self._generic_plotting_checks(fig, ax, nrows, ncols)
```

File: fda/grid.py [hunk 18]
```diff
@@ -677,92 +685,196 @@ def concatenate(self, other):
...
-        if ax is None:
-            ax = matplotlib.pyplot.gca()
-        _plot = None
-        for i in range(self.nsamples):
-            _plot = ax.scatter(self.sample_points[0],
-                               self.data_matrix[i],
-                               **kwargs)
-        self._set_labels(ax)
-        return _plot
+        fig, ax = self._genereic_plotting_checks(fig, ax, nrows, ncols)
```
Comment (vnmabus) [inline fda/grid.py]:
```suggestion
        fig, ax = self._generic_plotting_checks(fig, ax, nrows, ncols)
```

File: fda/grid.py [hunk 19]
```diff
@@ -677,92 +685,196 @@ def concatenate(self, other):
...
+            raise ValueError("fig and axes parameters cannot be passed as arguments at the same time.")
+
+        if fig is not None and len(fig.get_axes()) != self.ndim_image:
+            raise ValueError("Number of axes of the figure must be equal to"
+                             "the dimension of the image.")
+
+        if ax is not None and len(ax)!= self.ndim_image:
+            raise ValueError("Number of axes must be equal to the dimension of the image.")
+
+        if ax is not None and fig is not None and (nrows is not None or ncols is not None):
```
Comment (vnmabus) [inline fda/grid.py]: Maybe:

```suggestion
        if (ax is not None or fig is not None) and (nrows is not None or ncols is not None):
```

File: fda/grid.py [hunk 20]
```diff
@@ -677,92 +685,196 @@ def concatenate(self, other):
...
+            raise ValueError("The number of columns and/or number of rows of the figure, "
+                             "in which each dimension of the image is plotted, can only "
+                             "be customized in case fig is None and ax is None.")
+
+        if (nrows is not None and ncols is None) or (nrows is None and ncols is not None):
```
Comment (vnmabus) [inline fda/grid.py]: Why? We can use for the other value the minimum that works:
`ncols = math.ceil(self.ndim_image / nrows)`

File: fda/grid.py (hunk 15 above)
Comment (amandaher) [reply fda/grid.py]: changed

Comment (amandaher) [reply fda/grid.py]: changed

**Pull request merged**